### DISCUSSION TITLE
is BertFullTokenizer thread-safe?

### DISCUSSION THREAD

**Discussion by siddhsql** [deepjavalibrary/djl#3169] [Q&A]
is [this](https://github.com/deepjavalibrary/djl/blob/master/api/src/main/java/ai/djl/modality/nlp/bert/BertFullTokenizer.java) class thread-safe? it looks like so but i'd like to confirm.

**Comment by frankfliu** [comment]
I would recommend you to use Huggingface tokenizer instead. see: https://github.com/deepjavalibrary/djl/tree/master/extensions/tokenizers

**Comment by siddhsql** [comment]
thanks. could you explain the reason why?

**Comment by frankfliu** [comment]
1. DJL's `HuggingfaceTokenizer` has exactly the same behavior that matches Huggingface model
2. We can automatically import those tokenizers and model specific configurations from Huggingface hub
3. BertFullTokenizer doesn't have all the configurations compare to HuggingfaceTokenizer

**Comment by siddhsql** [comment]
how do we utilize the HF tokenizers which are written in Rust? asking for
my understanding.

**Comment by frankfliu** [comment]
[HuggingfaceTokenizer](https://javadoc.io/doc/ai.djl.huggingface/tokenizers/latest/ai/djl/huggingface/tokenizers/HuggingFaceTokenizer.html) is a DJL class

```
HuggingFaceTokenizer tokenizer = HuggingFaceTokenizer.newInstance("bert-base-cased");
Encoding encoding = tokenizer.encode(inputs);
```

See more exampels:
https://github.com/deepjavalibrary/djl/blob/master/extensions/tokenizers/src/test/java/ai/djl/huggingface/tokenizers/HuggingFaceTokenizerTest.java#L34

**Comment by siddhsql** [comment]
thanks. i guess my question is that did someone write Java equivalent of
the rust code or is it interoping with Rust somehow - how?

**Comment by frankfliu** [comment]
DJL uses JNI to interop with Rust.

**Comment by siddhsql** [comment]
thanks. one more question: where is the native library stored? is it packaged within some jar?

**Reply by frankfliu** [reply]
It packaged in this jar: https://mvnrepository.com/artifact/ai.djl.huggingface/tokenizers
